typing/typecore.ml

```diff
@@ -5144,6 +5350,121 @@ and type_function
+            end;
+            (Some (id, ety), (path, fl))
+        | Some (id, (path', fl'), ety), None ->
+            if !Clflags.principal
+                && get_level ty_expected < Btype.generic_level
+            then Location.prerr_warning loc
+                  (not_principal "this module unpacking");
+            (Some (id, ety), (path', fl'))
+      in
+      !check_package_closed ~loc:name.loc ~env
+                           ~typ:(newty (Tpackage (path, fl))) fl;
+      let mty = !Ctype.modtype_of_package env pparam_loc path fl in
+      let pv_uid = Uid.mk ~current_unit:(Env.get_current_unit ()) in
+      let arg_md = {
+        md_type = mty;
+        md_attributes = [];
+        md_loc = pparam_loc;
+        md_uid = pv_uid;
+      } in
+      let (res_ty, params, body, newtypes, contains_gadt), s_ident =
+        with_local_level begin fun () ->
+          let s_ident =
+            Ident.create_scoped ~scope:(Ctype.get_current_level()) name.txt
+          in
+          let new_env = Env.add_module_declaration ~check:true s_ident
+                              Mp_present arg_md env in
+          let expected_res = match id_expected_typ_opt with
+            | Some (id, ety) ->
+              let topt = instance_funct ~id_in:(Ident.of_unscoped id)
+                                        ~p_out:(Pident s_ident)
+                                        ~fixed:false ety in
+              begin match topt with Some t -> t | None -> ety end
+            | None -> newvar ()
+          in
+          type_function new_env rest body_constraint body
+              expected_res ~first:false ~in_function,
+          s_ident
+      end
+      in
+      let ident = Ident.Unscoped.create name.txt in
+      let exp_type =
+        match instance_funct ~p_out:(Pident (Ident.of_unscoped ident))
+                             ~id_in:s_ident ~fixed:false res_ty
+        with
+        | Some res_ty ->
+            Btype.newgenty (Tfunctor (arg_label, ident, (path, fl), res_ty))
+        | None ->
+            let pck_ty = newgenty (Tpackage (path, fl)) in
+            newgenty (Tarrow (arg_label, pck_ty, res_ty, commu_ok))
+      in
+      let _ =
+        try
+          unify env ty_expected exp_type
+        with Unify trace ->
+          raise (Error(loc, env, Expr_type_clash(trace, None, None)))
+      in
+      let pat_desc = Tpat_var (s_ident, name, pv_uid) in
+      let pattern = {
+        pat_desc;
+        pat_loc = pparam_loc;
+        pat_extra = [Tpat_unpack, pparam_loc, []];
+        pat_type = newty (Tpackage (path, fl));
+        pat_env = env;
+        pat_attributes = []
+      } in
+      let fp_kind = Tparam_pat pattern in
+      let param =
+        { fp_kind;
+          fp_arg_label = arg_label;
+          fp_param = s_ident;
+          fp_partial = Total;
+          fp_newtypes = newtypes;
+          fp_loc = pparam_loc;
+        }
+      in
+      exp_type, param :: params, body, [], contains_gadt
```

Comment:
Same here, there is a lot of code and I wish some were moved to an auxiliary function.